content/automate/ManageIQ/Service/Retirement/StateMachines/Methods.class/__methods__/get_retirement_entrypoint.rb

```diff
@@ -1,15 +1,59 @@
-service = $evm.root['service']
-if service.nil?
-  $evm.log('error', 'get_retirement_entrypoint: missing service object')
-  exit MIQ_ABORT
-end
+# frozen_string_literal: true
+
+#
+# Description: Resolves Service retirement entry point if present
+#
+
+module ManageIQ
+  module Automate
+    module Service
+      module Retirement
+        module StateMachines
+          module Methods
+            class GetRetirementEntrypoint
+              def initialize(handle = $evm)
+                @handle = handle
+              end
+
+              def main
+                retirement_entrypoint
+              end
+
+              private
+
+              def retirement_entrypoint
+                if service
+                  @handle.root['retirement_entry_point'] = entry_point
+                  @handle.root['service_action'] = 'Retirement'
+                  @handle.log("info", "retirement_entrypoint: #{entry_point}")
+                end
+              end
+
+              def service
+                @service ||= begin
```

Comment:
With the change to remove `if service` in `retirement_entrypoint ` there is no need to cache the results since the `service` method is only called once.

You could simplify to:
```ruby
def service
  if @handle.root['service'].nil?
    @handle.log('error', 'retirement_entrypoint: missing service object')
    exit MIQ_ABORT
  end

  @handle.root['service']
end
```

or even a `tap` version like this:
```ruby
def service
  @handle.root['service'].tap do |svc|
    if svc.nil?
      @handle.log('error', 'retirement_entrypoint: missing service object')
      exit MIQ_ABORT
    end
  end
end
```

Reply:
Or it can be simplified to:
```
def service
  @handle.root['service'].tap do |svc|
    raise 'retirement_entrypoint: missing service object'  if svc.nil?
  end
end
```

Reply:
@lfu Will the `raise` be processed the same as `exit MIQ_ABORT`?  cc @tinaafitz

Reply:
@gmcculloug @lfu .tap is roughly 50% slower in run time, and we do need to abort when it is nil, no? I am all for refactoring, but we should not be adding to run time when possible.

Reply:
@GregP I would not mind see the timing output from those tests as I expected we are talking about a minimal difference.

My suggestion, and in the spirit of refactoring, would be to do this:
```ruby
def service
  svc = @handle.root['service']
  if svc.nil?
    @handle.log('error', 'retirement_entrypoint: missing service object')
    exit MIQ_ABORT
  end

  svc
end
```